report/report-generators/src/main/kotlin/com/testerum/report_generators/reports/remote_server/RemoteServerExecutionListener.kt

```diff
@@ -0,0 +1,55 @@
+package com.testerum.report_generators.reports.remote_server
+
+import com.testerum.common_httpclient.HttpClientService
+import com.testerum.common_httpclient.TesterumHttpClientFactory
+import com.testerum.model.resources.http.request.HttpRequest
+import com.testerum.model.resources.http.request.HttpRequestBody
+import com.testerum.model.resources.http.request.enums.HttpRequestBodyType
+import com.testerum.model.resources.http.request.enums.HttpRequestMethod
+import com.testerum.report_generators.reports.utils.EXECUTION_LISTENERS_OBJECT_MAPPER
+import com.testerum.runner.cmdline.report_type.RunnerReportType
+import com.testerum.runner.cmdline.report_type.builder.EventListenerProperties.JsonEvents.REPORT_SERVER_URL
+import com.testerum.runner.events.execution_listener.ExecutionListener
+import com.testerum.runner.events.model.RunnerEvent
+
+class RemoteServerExecutionListener(private val properties: Map<String, String>) : ExecutionListener {
+
+    private val eventsAsStringBuilder = StringBuilder()
+
+    private val reportServerUrl = properties[REPORT_SERVER_URL]
+        ?: throw RuntimeException("${RunnerReportType.REMOTE_SERVER.toString()} requires the parameter \"${REPORT_SERVER_URL}\"")
+
+    override fun start() {}
+
+    override fun onEvent(event: RunnerEvent) {
+        if (eventsAsStringBuilder.isNotEmpty()) {
+            eventsAsStringBuilder.appendln()
+        }
+
+        eventsAsStringBuilder
+            .append(EXECUTION_LISTENERS_OBJECT_MAPPER.writeValueAsString(event))
+    }
+
+    override fun stop() {
+
+        TesterumHttpClientFactory.createHttpClient().use { httpClient ->
+            val httpClientService = HttpClientService(httpClient)
+
+            val request = HttpRequest(
+                method = HttpRequestMethod.POST,
+                url = "$reportServerUrl/report/v1",
+                body = HttpRequestBody(HttpRequestBodyType.RAW, eventsAsStringBuilder.toString()),
+                followRedirects = true
+            )
+            val response = httpClientService.executeHttpRequest(request)
+
+            if (400 <= response.statusCode) {
+                throw RuntimeException(
+                    """Exception appeared while calling the Report Server at URL: ${reportServerUrl}
+                   HTTP Request: ${request}
+                   HTTP Response: ${response}"""
```

Comment:
I'm not convinced that the request and the response have good ``toString`` methods.
If not, we could use the following:

```java
package com.fedex.o2p.http_client.utils.http;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.concurrent.Callable;
import lombok.NonNull;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.HexDump;
import org.apache.http.Header;
import org.apache.http.HttpEntity;
import org.apache.http.HttpEntityEnclosingRequest;
import org.apache.http.HttpRequest;
import org.apache.http.HttpResponse;
import org.apache.http.StatusLine;
import org.apache.http.util.EntityUtils;
import org.jetbrains.annotations.Nullable;

import static java.util.stream.Collectors.toUnmodifiableSet;

@Slf4j
public final class HttpClientDebugUtils {

    private static final Set<String> SENSITIVE_HEADERS =
            List.of(
                    "Authorization",
                    "Proxy-Authorization",
                    "X-Forwarded-For",
                    "X-Forwarded-Host",
                    "X-Csrf-Token",
                    "X-CSRFToken",
                    "Cookie",
                    "Set-Cookie",
                    "X-Request-Token"
            )
                .stream()
                .map(String::toLowerCase)
                .collect(toUnmodifiableSet());

    private static final int BINARY_SNIPPET_MAX_SIZE = 50 * 1024;

    private HttpClientDebugUtils() { }

    @NonNull
    public static String createMessageForRequest(@NonNull final HttpRequest request) {
        final var result = new StringBuilder();

        appendRequest(result, request);

        return result.toString();
    }

    @NonNull
    public static String createMessageForResponse(@NonNull final HttpResponse response) {
        final var result = new StringBuilder();

        appendResponse(result, response);

        return result.toString();
    }

    @NonNull
    public static String createMessage(@NonNull final String prefix,
                                       @NonNull final HttpRequest request,
                                       @NonNull final HttpResponse response) {
        final var result = new StringBuilder();

        result.append(prefix).append(":\n");

        appendRequest(result, request);
        result.append("\n\n");
        appendResponse(result, response);

        return result.toString();
    }

    @SneakyThrows
    private static void appendRequest(@NonNull final StringBuilder result,
                                      @NonNull final HttpRequest request) {
        result.append("Request:\n");
        result.append("--------\n");
        result.append(request.getRequestLine()).append('\n');
        result.append('\n');
        appendHeaders(result, request.getAllHeaders());
        result.append('\n');

        if (request instanceof HttpEntityEnclosingRequest) {
            final HttpEntityEnclosingRequest requestWithBody = (HttpEntityEnclosingRequest) request;

            result.append(getBodyForDebug(requestWithBody.getEntity()));
            result.append('\n');
        }
    }

    private static void appendResponse(@NonNull final StringBuilder result,
                                       @NonNull final HttpResponse response) {
        final StatusLine statusLine = response.getStatusLine();

        result.append("Response:\n");
        result.append("---------\n");
        result.append(statusLine.getStatusCode()).append(" ").append(statusLine.getReasonPhrase()).append('\n');
        result.append('\n');
        appendHeaders(result, response.getAllHeaders());
        result.append('\n');

        result.append(getBodyForDebug(response.getEntity()));
        result.append('\n');
    }

    private static void appendHeaders(@NonNull final StringBuilder result,
                                      @NonNull final Header[] headers) {
        if (headers == null) {
            return;
        }

        for (final Header header : headers) {
            final String headerName = header.getName();
            final String headerValue = header.getValue();

            result.append(headerName);
            result.append(": ");

            appendHeaderValue(result, headerName, headerValue);

            result.append('\n');
        }
    }

    private static void appendHeaderValue(@NonNull final StringBuilder result,
                                          @NonNull final String headerName,
                                          @Nullable final String headerValue) {
        if (isSensitiveHeader(headerName)) {
            result.append("*****");
        } else {
            result.append(headerValue);
        }
    }

    private static boolean isSensitiveHeader(@NonNull final String headerName) {
        return SENSITIVE_HEADERS.contains(headerName.toLowerCase(Locale.US));
    }

    @NonNull
    private static String getBodyForDebug(@Nullable final HttpEntity entity) {
        try {
            if (entity == null) {
                return "{{{ null }}}";
            }
            if (ContentTypesUtils.hasTextContent(entity)) {
                return EntityUtils.toString(entity);
            } else {
                return getHexDumpSnippetForDebug(entity::getContent);
            }
        } catch (final Exception e) {
            return "{{{ failed to get content for debug : " + printException(e) + " }}}";
        }
    }

    @NonNull
    private static String getHexDumpSnippetForDebug(@NonNull final Callable<InputStream> inputStreamSupplier) {
        try (final InputStream inputStream = inputStreamSupplier.call()) {
            final byte[] buffer = new byte[BINARY_SNIPPET_MAX_SIZE];

            final int actualBytesRead = inputStream.read(buffer);
            if (actualBytesRead != -1) {
                final byte[] startOfBodyBytes = new byte[actualBytesRead];
                System.arraycopy(buffer, 0, startOfBodyBytes, 0, actualBytesRead);

                final ByteArrayOutputStream hexDumpStream = new ByteArrayOutputStream();
                HexDump.dump(startOfBodyBytes, 0, hexDumpStream, 0);

                return new String(hexDumpStream.toByteArray(), StandardCharsets.UTF_8);
            } else {
                return "";
            }
        } catch (final Exception e) {
            final String errorMessage = "{ERROR: failed to get start of input stream: " + e.getMessage() + "}";

            LOG.warn(errorMessage, e);

            return errorMessage;
        }
    }

    @NonNull
    private static String printException(@NonNull final Throwable exception) {
        final StringWriter out = new StringWriter();

        exception.printStackTrace(
                new PrintWriter(out)
        );

        return out.toString();
    }

}
```

Reply:
This HttpRequest and HttpResponse is a Testerum object that is a data class.   By default has a an toString based on the properties, but probably is not that nicely formated like what you have in this class. 
The problem is that this class is based on HttpClinet. Would take some time to write it just for formatting.
If you think is something that make sens I will implemented it.